File: valid-palindrome/flynn.cpp
/**
 * 풀이
 * - 주어진 string `s`의 양 끝에서부터 차례대로 비교해가며 palindrome 여부를 판단합니다
 * 
 * Big-O
 * - N: 주어진 string `s`의 길이
 * 
 * - Time Complexity: O(N)
 *   - `s`가 palindrome인 경우, `s` 전체를 탐색하므로 O(N)의 시간복잡도를 가집니다
 * 
 * - Space Complexity: O(1)
 *   - 입력값과 무관하게 일정한 저장공간을 사용합니다
 */

class Solution {
public:
    bool isPalindrome(string s) {
        string::iterator lo = s.begin();
        string::iterator hi = s.end();

        while (lo <= hi) {
            if (isalnum(*lo) && isalnum(*hi)) {
                if (tolower(*lo) != tolower(*hi)) return false;
                else {
                    ++lo;
                    --hi;
                    continue;
                }
            }
            if (!isalnum(*lo)) ++lo;
            if (!isalnum(*hi)) --hi;
        }

        return true;
    }
};

File: maximum-depth-of-binary-tree/flynn.go
/*
풀이
- DFS
Big O
- N: 노드의 개수
- H: 트리의 높이
- Time complexity: O(N)
  - 모든 노드를 탐색합니다
- Space complexity: O(H)
  - 재귀호출 스택의 크기는 트리의 높이에 비례하여 증가합니다
*/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func maxDepth(root *TreeNode) int {
	maxDepth := 0
	var dig func(*TreeNode, int)
	dig = func(node *TreeNode, depth int) {
		if node == nil {
			if maxDepth < depth {
				maxDepth = depth
			}
			return
		}
		dig(node.Left, depth+1)
		dig(node.Right, depth+1)
	}
	dig(root, 0)
	return maxDepth
}

/*
풀이
- BFS
Big O
- N: 노드의 개수
- Time complexity: O(N)
  - 모든 노드를 탐색합니다
- Space complexity: O(N)
  - 노드 N개 짜리 트리에서 한 층의 폭은 N을 넘지 않습니다
    따라서 queue의 공간복잡도는 O(N)입니다
*/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
	level := 0
	queue := make([]*TreeNode, 0)
	if root != nil {
		queue = append(queue, root)
		level++
	}
	for len(queue) > 0 {
		currQSize := len(queue)
		for currQSize > 0 {
			node := queue[0]
			queue = queue[1:]
			currQSize--
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		if len(queue) > 0 {
			level++
		}
	}
	return level
}
